context/FormProvider.tsx

```diff
@@ -0,0 +1,158 @@
+import { useAccount, useContractRead } from "@starknet-react/core";
+import React, { FunctionComponent, useState } from "react";
+import { createContext, useMemo } from "react";
+import { computeMetadataHash, generateSalt } from "@/utils/userDataService";
+import { useNftPpVerifierContract } from "@/hooks/contracts";
+import { Abi } from "starknet";
+import { filterAssets, retrieveAssets } from "@/utils/nftService";
+
+interface FormState {
+  email: string;
+  isSwissResident: boolean;
+  tokenId: number;
+  duration: number;
+  selectedDomains: Record<string, boolean>;
+  // metadata
+  salt?: string;
+  metadataHash?: string;
+  needMetadata: boolean;
+  // tax
+  salesTaxRate: number;
+  //pfp
+  selectedPfp?: StarkscanNftProps;
+  //upsell
+  isUpselled: boolean;
+}
+
+type FormConfig = {
+  formState: FormState;
+  clearForm: () => void;
+  updateFormState: (updates: Partial<FormState>) => void;
+  userNft?: StarkscanNftProps[];
+};
+
+const initialState: FormState = {
+  email: "",
+  isSwissResident: false,
+  tokenId: 0,
+  duration: 1,
+  selectedDomains: {},
+  needMetadata: false,
+  salesTaxRate: 0,
+  selectedPfp: undefined,
+  isUpselled: false,
+};
+
+export const FormContext = createContext<FormConfig>({
+  formState: initialState,
+  clearForm: () => {},
+  updateFormState: () => {},
+  userNft: [],
+});
+
+export const FormProvider: FunctionComponent<Context> = ({ children }) => {
+  const { address } = useAccount();
+  const [formState, setFormState] = useState<FormState>(initialState);
+  const [userNft, setUserNft] = useState<StarkscanNftProps[]>([]);
+  // get whitelisted NFT contract addresses
+  const { contract } = useNftPpVerifierContract();
+  const { data: whitelistData, error: whitelistError } = useContractRead({
+    address: contract?.address as string,
+    abi: contract?.abi as Abi,
+    functionName: "get_whitelisted_contracts",
+    args: [],
+  });
+
+  const updateFormState = (updates: Partial<FormState>) => {
+    setFormState((prevState) => ({ ...prevState, ...updates }));
+  };
+
+  const clearForm = () => {
+    setFormState((prevState) => ({
+      ...prevState,
+      selectedDomains: {},
+      tokenId: 0,
+      duration: 1,
+      isUpselled: false,
+      selectedPfp: undefined,
+    }));
+  };
+
+  // Check if connected account has whitelisted NFTs
+  // We fetch nfts once for all the app
+  useMemo(() => {
+    if (!address || !whitelistData) {
+      setUserNft([]);
+      return;
+    }
+    retrieveAssets(
+      `${process.env.NEXT_PUBLIC_SERVER_LINK}/starkscan/fetch_nfts`,
+      address
+    )
+      .then((data) => {
+        setUserNft(filterAssets(data.data, whitelistData as bigint[]));
+      })
+      .catch(() => {
+        setUserNft([]);
```

Comment:
So if starkscan is down or we're blacklisted for some reason. 

We just don't make people set NFTs right ?

Reply:
yes